### PR TITLE
Add Application Connectivity Validator - migration to Istio

### PR DESCRIPTION
**Pull request by Szymongib**
**Description**

While using Istio as Ingress for Application Connector we need a way to validate Subject of the Client certificate. New component Application Connectivity Validator is created for that purpose. 

Changes proposed in this pull request:

- Add new component - Application Connectivity Validator
- Use Virtual Service instead of Ingress for Applications

**Related issue(s)**
#1608

### CODE REVIEW

File: components/application-connectivity-validator/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+- **eventServicePathPrefix** - Path prefix for which requests will be forwarded to the Event Service. The default value is `/v1/events`.
+- **eventServiceHost** - Host (and port) of the Event Service. The default value is `events-api:8080`.
+- **appRegistryPathPrefix** - Path prefix for which requests will be forwarded to the Application Registry. The default value is `/v1/metadata`.
+- **appRegistryHost** - Host (and port) of the Event Service. The default value is `application-registry-external-api:8081`.
+
+
+## Details
+
+The validation is based on `X-Forwarded-Client-Cert` header. 
+The header is added to the request by the Envoy Proxy, after successful client certificate validation, which is configured in the Istio Gateway.
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
After successful verification, the Envoy Proxy adds the header to the request. The conditions of successful verification are defined in the Istio Gateway.
```

File: components/application-connectivity-validator/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+For the header to be passed down to the service the mutual TLS between Istio sidecars have to be enabled.
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
The service to which the header is added must have mutual TLS between Istio sidecar Pods enabled. 
```

File: components/application-connectivity-validator/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+Example `X-Forwarded-Client-Cert` header has the following format:
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
This is an example `X-Forwarded-Client-Cert` header:
```

File: components/application-connectivity-validator/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+`Hash=f4cf22fb633d4df500e371daf703d4b4d14a0ea9d69cd631f95f9e6ba840f8ad;Subject="CN=test-application,OU=OrgUnit,O=Organization,L=Waldorf,ST=Waldorf,C=DE";URI=,By=spiffe://cluster.local/ns/kyma-integration/sa/default;Hash=6d1f9f3a6ac94ff925841aeb9c15bb3323014e3da2c224ea7697698acf413226;Subject="";URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account`
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]: Please make this a code block - use three backticks instead of one.

File: components/application-connectivity-validator/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+Multiple certificate information is due to the client certificate used in mutual TLS communication between sidecars.
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
The certificate contains information about multiple certificates because of the client certificate used in mTLS-secure communication between sidecars of a service.
```

File: components/application-connectivity-validator/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+
+The Application Connectivity Validator will forward only requests which contain the `X-Forwarded-Client-Cert` header containing `Subject` with the following fields corresponding to the Application Custom Resource:
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
The Application Connectivity Validator forwards only the requests that contain the `X-Forwarded-Client-Cert` header that contains `Subject` with the following fields corresponding to the Application custom resource:
```

File: components/application-connectivity-validator/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+- **CommonName** - Application Custom Resource name
```
Comment (tomekpapiernik) [inline components/application-connectivity-validator/README.md]:
```suggestion
- **CommonName** - Application custom resource name
```

File: components/application-connectivity-validator/internal/validationproxy/handler.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+		return
+	}
+
+	host, err := ph.determineHost(r.URL.Path)
+	if err != nil {
+		httptools.RespondWithError(w, err)
+		return
+	}
+
+	reverseProxy := &httputil.ReverseProxy{
```
Comment (akgalwas) [inline components/application-connectivity-validator/internal/validationproxy/handler.go]: I would suggest to cache ReverseProxy objects here.

Comment (Szymongib) [reply components/application-connectivity-validator/internal/validationproxy/handler.go]: As discussed, we actually need only two instances of ReverseProxy, therefore there is no need for caching.

File: components/application-connectivity-validator/internal/validationproxy/handler_test.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+			tenant:          tenant,
+			group:           group,
+			certInfoHeader: `Hash=f4cf22fb633d4df500e371daf703d4b4d14a0ea9d69cd631f95f9e6ba840f8ad;Subject="CN=test-application,OU=group,O=tenant,L=Waldorf,ST=Waldorf,C=DE";` +
+				`URI=,By=spiffe://cluster.local/ns/kyma-integration/sa/default;` +
+				`Hash=6d1f9f3a6ac94ff925841aeb9c15bb3323014e3da2c224ea7697698acf413226;Subject="";` +
+				`URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account`,
+			expectedStatus: http.StatusOK,
+		},
+		{
+			caseDescription: "Application with group and tenant and Common Name is invalid",
```
Comment (akgalwas) [inline components/application-connectivity-validator/internal/validationproxy/handler_test.go]: I would also consider changing a description of this use case to something like : "Application in the central context with invalid Common Name"

File: components/application-connectivity-validator/internal/validationproxy/handler_test.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+		{
+			caseDescription: "Application without group and tenant",
+			certInfoHeader: `Hash=f4cf22fb633d4df500e371daf703d4b4d14a0ea9d69cd631f95f9e6ba840f8ad;Subject="CN=test-application,OU=OrgUnit,O=Organization,L=Waldorf,ST=Waldorf,C=DE";` +
+				`URI=,By=spiffe://cluster.local/ns/kyma-integration/sa/default;` +
+				`Hash=6d1f9f3a6ac94ff925841aeb9c15bb3323014e3da2c224ea7697698acf413226;Subject="";` +
+				`URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account`,
+			expectedStatus: http.StatusOK,
+		},
+		{
+			caseDescription: "Application without group and tenant and Common Name is invalid",
```
Comment (akgalwas) [inline components/application-connectivity-validator/internal/validationproxy/handler_test.go]: I would also consider changing a description of this use case to something like : "Application in the local context with invalid Common Name"

File: components/application-connectivity-validator/internal/validationproxy/handler_test.go [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+			caseDescription: "Application with group and tenant and Common Name is invalid",
+			tenant:          tenant,
+			group:           group,
+			certInfoHeader: `Hash=f4cf22fb633d4df500e371daf703d4b4d14a0ea9d69cd631f95f9e6ba840f8ad;Subject="CN=invalid-application,OU=group,O=tenant,L=Waldorf,ST=Waldorf,C=DE";` +
+				`URI=,By=spiffe://cluster.local/ns/kyma-integration/sa/default;` +
+				`Hash=6d1f9f3a6ac94ff925841aeb9c15bb3323014e3da2c224ea7697698acf413226;Subject="";` +
+				`URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account`,
+			expectedStatus: http.StatusForbidden,
+		},
+		{
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/application-connectivity-validator/internal/validationproxy/handler_test.go]: Good catch

File: components/application-connectivity-validator/internal/validationproxy/handler_test.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+			caseDescription: "Application with group and tenant and Organization is invalid",
+			tenant:          tenant,
+			group:           group,
+			certInfoHeader: `Hash=f4cf22fb633d4df500e371daf703d4b4d14a0ea9d69cd631f95f9e6ba840f8ad;Subject="CN=test-application,OU=,O=tenant,L=Waldorf,ST=Waldorf,C=DE";` +
+				`URI=,By=spiffe://cluster.local/ns/kyma-integration/sa/default;` +
+				`Hash=6d1f9f3a6ac94ff925841aeb9c15bb3323014e3da2c224ea7697698acf413226;Subject="";` +
+				`URI=spiffe://cluster.local/ns/istio-system/sa/istio-ingressgateway-service-account`,
+			expectedStatus: http.StatusForbidden,
+		},
+		{
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply components/application-connectivity-validator/internal/validationproxy/handler_test.go]: Done

File: components/application-connectivity-validator/internal/validationproxy/handler_test.go (hunk 9 above)
Comment (Szymongib) [reply components/application-connectivity-validator/internal/validationproxy/handler_test.go]: I am not sure about that. I think that would leak some information to that component that is not necessarily required in its context.
What do you think?

Comment (akgalwas) [reply components/application-connectivity-validator/internal/validationproxy/handler_test.go]: Ok, we can leave it as it is but I would slightly rephrase it to increase readability (e.g. "Application with group, tenant and invalid Common Name")

Comment (Szymongib) [reply components/application-connectivity-validator/internal/validationproxy/handler_test.go]: Sure, I think it is a good idea.
